src/ws/hooks/app.ts

```diff
@@ -0,0 +1,233 @@
+/**
+ * Graasp websocket client
+ * React effect hooks to subscribe to real-time updates and mutate query client
+ */
+import { List } from 'immutable';
+import { useEffect } from 'react';
+
+import { UUID, convertJs } from '@graasp/sdk';
+import { AppActionRecord, AppDataRecord, AppSettingRecord } from '@graasp/sdk/frontend';
+
+import { Channel, WebsocketClient } from '../ws-client';
+import { useQueryClient } from '@tanstack/react-query';
+import { buildAppActionsKey, buildAppDataKey, buildAppSettingsKey } from '../../config/keys';
+import {
+  AppActionEvent,
+  AppDataEvent,
+  AppEventKinds,
+  AppOperations,
+  AppSettingEvent,
+} from '../types';
+import { APP_ACTIONS_TOPIC, APP_DATA_TOPIC, APP_SETTINGS_TOPIC } from '../constants';
+
+export const configureWsAppDataHooks = (websocketClient?: WebsocketClient) => {
+  if (!websocketClient) {
+    return {
+      useAppDataUpdates: () => {
+        // do nothing
+        console.warn('No websocket client was provided.');
+      },
+    };
+  }
```

Comment:
I see what you did there 😛 I guess you are technically correct that this does not break the rules of hooks. However, IMO it's less readable and maintanable to have this "no-op hook" that stubs the implementation when the ws client is not defined.

Have a look at the way item ws updates are implemented:
- https://github.com/graasp/graasp-query-client/blob/547495b6e5381f44e99a21f51389907777b41e5d/src/hooks/item.ts#L258-L282
- https://github.com/graasp/graasp-query-client/blob/547495b6e5381f44e99a21f51389907777b41e5d/src/ws/hooks/item.ts#L33-L75

Technically, you would be right in saying that there is a conditional on `itemWsHooks?.useItemUpdates(getUpdates ? id : null);` so this is breaking the rules of hook. However note that the nullable condition only depends on `const itemWsHooks = enableWebsocket && websocketClient` and they are both "static" configuration, in the sense that they do not change after initializing the entire front-end application. So we can guarantee that the condition is stable: in the entire lifecycle of the app, either the hooks are configured and exist, or they do not. Note that in your previous implementation you had:
```ts
      if (
        typeof useAppActionsUpdates !== 'undefined' &&
        getUpdates &&
        permissionLevel === PermissionLevel.Admin
      ) {
        useAppActionsUpdates(itemId);
      }
```
in which case both `getUpdates` and the `permissionLevel` checks are dynamic at runtime (e.g. they may change during the lifecycle of the application at the different call sites) so in this case the rules of hook can be broken at runtime without warning.

WDYT @spaenleh @swouf @pyphilia ? I think it's worth having `wsAppDataHooks?.useAppDataUpdates()` at call site if it only depends on the global static configuration of `queryConfig.enableWebsocket && queryConfig.websocketClient` because the code readibility is improved. Or we can decide that the rules of hooks should be absolute, and hence implement it as in the current code, with the no-op hook.

So in any case, we should decide on one convention and stick to it for consistency, so we either change the app updates, or the item updates.

Reply:
Update: after discussion we can actually both respect the rules of hooks and reduce the boiler plate, by simply considering the `enableWebsocket && websocketClient` from within the deepest hook itself instead of at configuration or call time. Since we can capture the configuration value from within the closure of the hook function, there is no need for additional checks at call site.

I.e.

```ts
// this is wrong
const wsHooks = enableWebsocket && websocketClient ? configureFooHooks() : undefined;
useFoo() {
    wsHooks?.useFooUpdates() // <-- ! rules of hooks
}
configureFooHooks() {
   return {
     useFooUpdates: () => { ... implementation }
   }
}

// this is verbose
const { useFooUpdates } = configureFooHooks()
useFoo() {
   useFooUpdates()
}
configureFooHooks(client) {
   if (!client) return {
     useFooUpdates: () => { /* no-op hook */ }
   } else return {
     useFooUpdates: () => { ... implementation }
   }
}

// this is correct
const { useFooUpdates } = configureFooHooks()
useFoo() {
   useFooUpdates()
}
configureFooHooks(client) {
   return {
       useFooUpdates() { // <-- single definition
           useEffect() {
               if (!client || !wsEnabled) return;    // <-- call stack is preserved so no rules of hooks violation
               else // ... implementation
           }
           ...
       }
   }
}
```